### PR TITLE
codeartifact resources tagging support

### LINKED ISSUE #15958: AWS CodeArtifact resource tagging
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

CodeArtifact domains and repositories can now be tagged.

### New or Affected Resource(s)

* [`aws_codeartifact_domain`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/codeartifact_domain)
* [`aws_codeartifact_repository`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/codeartifact_repository)

### Potential Terraform Configuration

```hcl
resource "aws_codeartifact_domain" "example" {
  tags = {
    Environment = "test"
  }
}

resource "aws_codeartifact_repository" "example" {
  tags = {
    Environment = "test"
  }
}
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2020/11/aws-codeartifact-now-supports-resource-tagging/).
[User Guide](https://docs.aws.amazon.com/codeartifact/latest/ug/tag-resources.html).

Requires [AWS SDK v1.35.18](https://github.com/aws/aws-sdk-go/releases/tag/v1.35.18):

* https://github.com/terraform-providers/terraform-provider-aws/pull/15930

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #15958

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_codeartifact_domain - add `tags` argument.
resource_aws_codeartifact_repository - add `tags` argument.
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'
--- PASS: TestAccAWSCodeArtifactDomain_disappears (46.43s)
--- PASS: TestAccAWSCodeArtifactDomain_basic (56.14s)
--- PASS: TestAccAWSCodeArtifactDomain_tags (142.23s)

--- PASS: TestAccAWSCodeArtifactRepository_disappears (55.96s)
--- PASS: TestAccAWSCodeArtifactRepository_basic (62.72s)
--- PASS: TestAccAWSCodeArtifactRepository_owner (65.24s)
--- PASS: TestAccAWSCodeArtifactRepository_description (111.50s)
--- PASS: TestAccAWSCodeArtifactRepository_tags (153.52s)
--- PASS: TestAccAWSCodeArtifactRepository_externalConnection (158.15s)
--- PASS: TestAccAWSCodeArtifactRepository_upstreams (178.47s)
```

### CODE REVIEW

File: aws/resource_aws_codeartifact_domain.go [hunk 1]
```diff
@@ -99,17 +104,41 @@ func resourceAwsCodeArtifactDomainRead(d *schema.ResourceData, meta interface{})
...
 	d.Set("encryption_key", sm.Domain.EncryptionKey)
 	d.Set("owner", sm.Domain.Owner)
 	d.Set("asset_size_bytes", sm.Domain.AssetSizeBytes)
 	d.Set("repository_count", sm.Domain.RepositoryCount)
 	d.Set("created_time", sm.Domain.CreatedTime.Format(time.RFC3339))
 
+	tags, err := keyvaluetags.CodeartifactListTags(conn, arn)
+
+	if err != nil {
+		return fmt.Errorf("error listing tags for Codeartifact Domain (%s): %w", arn, err)
```
Comment (ewbankkit) [inline aws/resource_aws_codeartifact_domain.go:119]:
```suggestion
		return fmt.Errorf("error listing tags for CodeArtifact Domain (%s): %w", arn, err)
```

File: aws/resource_aws_codeartifact_domain.go [hunk 2]
```diff
@@ -99,17 +104,41 @@ func resourceAwsCodeArtifactDomainRead(d *schema.ResourceData, meta interface{})
...
 	return nil
 }
 
+func resourceAwsCodeArtifactDomainUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codeartifactconn
+
+	if d.HasChange("tags") {
+		o, n := d.GetChange("tags")
+		if err := keyvaluetags.CodeartifactUpdateTags(conn, d.Get("arn").(string), o, n); err != nil {
+			return fmt.Errorf("error updating Codeartifact Domain (%s) tags: %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_codeartifact_domain.go:135]:
```suggestion
			return fmt.Errorf("error updating CodeArtifact Domain (%s) tags: %w", d.Id(), err)
```

File: aws/resource_aws_codeartifact_repository.go [hunk 3]
```diff
@@ -194,11 +202,19 @@ func resourceAwsCodeArtifactRepositoryUpdate(d *schema.ResourceData, meta interf
 		}
 	}
 
+	if d.HasChange("tags") {
+		o, n := d.GetChange("tags")
+		if err := keyvaluetags.CodeartifactUpdateTags(conn, d.Get("arn").(string), o, n); err != nil {
+			return fmt.Errorf("error updating Codeartifact Repository (%s) tags: %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_codeartifact_repository.go:208]:
```suggestion
			return fmt.Errorf("error updating CodeArtifact Repository (%s) tags: %w", d.Id(), err)
```

File: aws/resource_aws_codeartifact_repository.go [hunk 4]
```diff
@@ -239,6 +256,16 @@ func resourceAwsCodeArtifactRepositoryRead(d *schema.ResourceData, meta interfac
 		}
 	}
 
+	tags, err := keyvaluetags.CodeartifactListTags(conn, arn)
+
+	if err != nil {
+		return fmt.Errorf("error listing tags for Codeartifact Repository (%s): %w", arn, err)
```
Comment (ewbankkit) [inline aws/resource_aws_codeartifact_repository.go:262]:
```suggestion
		return fmt.Errorf("error listing tags for CodeArtifact Repository (%s): %w", arn, err)
```

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSCodeArtifactDomain_basic\|TestAccAWSCodeArtifactDomain_tags\|TestAccAWSCodeArtifactRepository_basic\|TestAccAWSCodeArtifactRepository_tags'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCodeArtifactDomain_basic\|TestAccAWSCodeArtifactDomain_tags\|TestAccAWSCodeArtifactRepository_basic\|TestAccAWSCodeArtifactRepository_tags -timeout 120m
=== RUN   TestAccAWSCodeArtifactDomain_basic
=== PAUSE TestAccAWSCodeArtifactDomain_basic
=== RUN   TestAccAWSCodeArtifactDomain_tags
=== PAUSE TestAccAWSCodeArtifactDomain_tags
=== RUN   TestAccAWSCodeArtifactRepository_basic
=== PAUSE TestAccAWSCodeArtifactRepository_basic
=== RUN   TestAccAWSCodeArtifactRepository_tags
=== PAUSE TestAccAWSCodeArtifactRepository_tags
=== CONT  TestAccAWSCodeArtifactDomain_basic
=== CONT  TestAccAWSCodeArtifactRepository_tags
=== CONT  TestAccAWSCodeArtifactRepository_basic
=== CONT  TestAccAWSCodeArtifactDomain_tags
--- PASS: TestAccAWSCodeArtifactDomain_basic (29.93s)
--- PASS: TestAccAWSCodeArtifactRepository_basic (32.68s)
--- PASS: TestAccAWSCodeArtifactDomain_tags (63.67s)
--- PASS: TestAccAWSCodeArtifactRepository_tags (70.49s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	70.552s
```

**Comment by breathingdust** [comment]
LGTM :rocket: Thanks @DrFaust92 

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSCodeArtifactDomain_basic\|TestAccAWSCodeArtifactDomain_tags\|TestAccAWSCodeArtifactRepository_basic\|TestAccAWSCodeArtifactRepository_tags'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCodeArtifactDomain_basic\|TestAccAWSCodeArtifactDomain_tags\|TestAccAWSCodeArtifactRepository_basic\|TestAccAWSCodeArtifactRepository_tags -timeout 120m
=== RUN   TestAccAWSCodeArtifactDomain_basic
=== PAUSE TestAccAWSCodeArtifactDomain_basic
=== RUN   TestAccAWSCodeArtifactDomain_tags
=== PAUSE TestAccAWSCodeArtifactDomain_tags
=== RUN   TestAccAWSCodeArtifactRepository_basic
=== PAUSE TestAccAWSCodeArtifactRepository_basic
=== RUN   TestAccAWSCodeArtifactRepository_tags
=== PAUSE TestAccAWSCodeArtifactRepository_tags
=== CONT  TestAccAWSCodeArtifactDomain_basic
=== CONT  TestAccAWSCodeArtifactRepository_tags
=== CONT  TestAccAWSCodeArtifactRepository_basic
=== CONT  TestAccAWSCodeArtifactDomain_tags
2020/11/04 13:52:28 [DEBUG] Waiting for state to become: [success]
--- PASS: TestAccAWSCodeArtifactDomain_basic (20.36s)
--- PASS: TestAccAWSCodeArtifactRepository_basic (22.66s)
--- PASS: TestAccAWSCodeArtifactDomain_tags (53.24s)
--- PASS: TestAccAWSCodeArtifactRepository_tags (54.53s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	56.089s
```

**Pull request merged**